Common SQLCheckOperators Various Functionality Update

denimalpaca opened the pull request:
The SQLCheckOperators under the Common provider have some issues since their initial release, including:
- Lack of batching
- A bug in the `SQLTableCheckOperator` that forces users to use only fully-aggregated checks
- A bug in the output of the test failure exception in the `SQLColumnCheckOperator` that only ever returns the final SQL query built

This PR remedies these issues.

closes: #25163

Commit:
Add batching to SQL Check Operators

Commit adds a WHERE clause to the sql statement that allows for
arbitrary batching in a given table.

Commit:
Add batching to SQL Check Operators

Commit adds a WHERE clause to the sql statement that allows for
arbitrary batching in a given table.

Commit:
Add batching to SQL Check Operators

Commit adds a WHERE clause to the sql statement that allows for
arbitrary batching in a given table.

Commit:
Fix bug with multiple table checks

When multiple table checks are given to the SQLTableCheckOperator
and at least one is not a fully aggregate statement, a GROUP BY
clause was previously needed. This commit updates the operator to
use the get_pandas_df() method instead of _get_first() to return a
pandas dataframe object that contains the check names and check
results from the new style of query. The new style of query uses
UNION ALL to run each test as its own SELECT statement, bypassing
the need to do a GROUP BY.

Commit:
Fix bug with multiple table checks

When multiple table checks are given to the SQLTableCheckOperator
and at least one is not a fully aggregate statement, a GROUP BY
clause was previously needed. This commit updates the operator to
use the get_pandas_df() method instead of _get_first() to return a
pandas dataframe object that contains the check names and check
results from the new style of query. The new style of query uses
UNION ALL to run each test as its own SELECT statement, bypassing
the need to do a GROUP BY.

Commit:
Fix bug with multiple table checks

When multiple table checks are given to the SQLTableCheckOperator
and at least one is not a fully aggregate statement, a GROUP BY
clause was previously needed. This commit updates the operator to
use the get_pandas_df() method instead of _get_first() to return a
pandas dataframe object that contains the check names and check
results from the new style of query. The new style of query uses
UNION ALL to run each test as its own SELECT statement, bypassing
the need to do a GROUP BY.

Commit:
Update test failure logic

Changed name of method from _get_failed_tests to _get_failed_checks
to better match naming, and updated logic of the method to include
an optional column param. The query in the column check operator
is removed from the failed test exception message, as it was only
ever showing the last query, instead of the relevant one(s). This is
replaced by the column, which will be more useful in debugging.

Commit:
Update test failure logic

Changed name of method from _get_failed_tests to _get_failed_checks
to better match naming, and updated logic of the method to include
an optional column param. The query in the column check operator
is removed from the failed test exception message, as it was only
ever showing the last query, instead of the relevant one(s). This is
replaced by the column, which will be more useful in debugging.

Commit:
Update test failure logic

Changed name of method from _get_failed_tests to _get_failed_checks
to better match naming, and updated logic of the method to include
an optional column param. The query in the column check operator
is removed from the failed test exception message, as it was only
ever showing the last query, instead of the relevant one(s). This is
replaced by the column, which will be more useful in debugging.

Commit:
Add table alias to SQLTableCheckOperator query

Without a table alias, the query does not run on Postgres and
other databases. The alias is arbitrary and used only for
proper query execution.

Commit:
Add table alias to SQLTableCheckOperator query

Without a table alias, the query does not run on Postgres and
other databases. The alias is arbitrary and used only for
proper query execution.

Commit:
Add table alias to SQLTableCheckOperator query

Without a table alias, the query does not run on Postgres and
other databases. The alias is arbitrary and used only for
proper query execution.

Commit:
Fix formatting error in operator

Commit:
Fix formatting error in operator

Commit:
Fix formatting error in operator

Commit:
Merge branch 'sql_check_operators_various_functionality_update' of github.com:denimalpaca/airflow into sql_check_operators_various_functionality_update

Commit:
Move alias to proper query build statement

The table alias should be in the self.sql query build statement
as that is where the table it needs to alias is defined.

eladkal commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -73,6 +79,7 @@ class SQLColumnCheckOperator(BaseSQLOperator):
             }
         }
 
+    :param batch: a SQL statement that is added to a WHERE clause to create batches
```
I'm not sure if this parameter name is self explained and even after reading the description I still dont understand how to to use it (I must check example/test/source code to understand)
Can we make it self explained?

potiuk replied:
Agree. I also would not have guessed what it is about

denimalpaca replied:
Pushed a change to clarify, please let me know if it makes more sense now

potiuk replied:
Hmm. I think a better name for that one would be `partition_clause`? Batch is rather associated with files but for SQL Tables, partitioning is likely more accurate name (Assuming that I correctly guessed what's the intended use).

denimalpaca replied:
Yep -- the idea here is to allow someone to not have to run a test on the full table, they could partition by date, or some other field. I clarified more in the docstring and added an e.g.

Commit:
Merge branch 'sql_check_operators_various_functionality_update' of github.com:denimalpaca/airflow into sql_check_operators_various_functionality_update

Commit:
Bug fixes and updates to test and operator

Fixed bug in test where the dataframe column names did not match
the operator's expected dataframe column names. Added more info
to the SQLColumnCheckOperator's batch arg. Fixed the location of
table aliasing in SQLTableCheckOperator.

Commit:
Remove merge conflict lines

Commit:
Rename parameter batch to partition_clause

Gives a clearer name to the parameter and adds templating to
the SQLTableCheckOperator.

Commit:
Fix typo in docstring

Commit:
Reformat operator file

potiuk reviewed (approved):
Now it looks great! Thanks @denimalpaca for being so responsive to our comments :)

ashb commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -273,38 +303,38 @@ def __init__(
         )
+        partition_clause_statement = f"WHERE {self.partition_clause}" if self.partition_clause else ""
+        self.sql = f"SELECT check_name, check_result FROM ({checks_sql}) "
+        f"AS check_table {partition_clause_statement};"
```
Whoops. Missing `+` here. This string is just a literal, not part of `self.sql` (and thus the entire `partition_clause_statement` isn't used!

denimalpaca replied:
Just got back from vacation, I see this got merged without a fix here. I can open a separate PR to address it.

potiuk replied:
Please.

potiuk replied:
Yeah. Ash (nick-name `Hawk-Eye`) noticed it after we merged and started voting on it. I decided to continue to release as both operators were new but I would love to get it fixed for 1.2.0.

I hope you had good vacation @denimalpaca :D

ashb commented on airflow/providers/common/sql/operators/sql.py:
```diff
@@ -273,38 +303,38 @@ def __init__(
 
-        self.sql = f"SELECT {check_mins_sql} FROM (SELECT {checks_sql} FROM {self.table});"
-        records = hook.get_first(self.sql)
+        records = hook.get_pandas_df(self.sql)
```
Why did we change from getting records to getting this as a pandas dataframe? This now places a _hard_ requirement on using pandas for this operator, where as previously pandas was almost entirely optional.

denimalpaca replied:
So, this was changed because with `hook.get_first`, there was an issue with how the SQL was being written that caused *only* fully aggregated checks to be returned, unless the syntax of the SQL query was changed, but that would require either a `fetch_all` or `get_pandas_df` call as the new SQL needs to returned multiple lines. It seemed much easier and possibly more efficient to use pandas here, but if a `fetch_all` seems more reasonable this can be changed. Happy to explain more about the specific issue if curious.

potiuk replied:
I'd be curious to hear more about it :D

denimalpaca replied:
To expand a bit more, a check like `col_a + col_b >= col_c` would not work when there were multiple checks in the operator as the previous `SELECT` statement would then fail and require a `GROUP BY` clause iirc. So the check would either have to be in its own operator, or be amended like so: `SUM(col_a) + SUM(col_b) >= SUM(col_c)` which isn't quite the same check. So the query needed to be updated, and the one that I wound up using returns multiple rows. So `get_first` is no longer useful, and in the moment of writing it seemed that handling things with a pandas dataframe might be easier in the long term, if more complicated uses of the pulled in data were implemented. But as of now I see how it's unneeded.

potiuk replied:
Thanks.... Is the operator easy to understand by the users? I am afraid this is something we need good and clear howto and examples for it because people will have hard time using it and rais too many questions.

denimalpaca replied:
Ideally most users won't need to learn about why `.fetch_all` is being used instead of `.get_first` 🙃 . I'm working with some users of the operator right now and seeing what's complicated to make sure the docs are robust. I also have a working example DAG showing how to use the operator (with several more planned) [here](https://registry.astronomer.io/providers/common-sql/modules/sqltablecheckoperator/#example-dags).
